Feature:

CarlosZ96 opened the pull request:
Learning to use transactions and filtering with operators

![rollbackspecies](https://github.com/CarlosZ96/first-look-at-sql/assets/125598459/d0d80630-77c0-4b28-994c-8f6693715e57)
![beforespecies](https://github.com/CarlosZ96/first-look-at-sql/assets/125598459/ccdc6fe8-02df-409d-b4b8-07a5698d7226)
![escapeavergepokemon](https://github.com/CarlosZ96/first-look-at-sql/assets/125598459/5ed8c767-ff36-4d08-9647-0d9e1b2fe633)
![escapeaverge](https://github.com/CarlosZ96/first-look-at-sql/assets/125598459/28c9a637-2f7f-4b2f-a73c-7e32bceeab4a)
![maxweight](https://github.com/CarlosZ96/first-look-at-sql/assets/125598459/e552ec4d-2978-44f9-99e4-a85c1b01e99e)
![minweight](https://github.com/CarlosZ96/first-look-at-sql/assets/125598459/16412952-c772-412e-a127-633a74576ca9)
![notneutered](https://github.com/CarlosZ96/first-look-at-sql/assets/125598459/d61cd31f-bf50-4806-b411-dcb3d8672d93)
![neuteredscapes](https://github.com/CarlosZ96/first-look-at-sql/assets/125598459/7da0fb13-0e4d-4a77-a9a8-76acd669978b)
![average](https://github.com/CarlosZ96/first-look-at-sql/assets/125598459/18890c6d-1152-400f-b972-73af87a33623)
![neverscape](https://github.com/CarlosZ96/first-look-at-sql/assets/125598459/c8d7db08-0d86-488e-bcde-7986d193a8c2)
![count](https://github.com/CarlosZ96/first-look-at-sql/assets/125598459/a41fce8b-bfa3-4f6a-870d-9ecffd31ee82)
![rollbackdateofbirth](https://github.com/CarlosZ96/first-look-at-sql/assets/125598459/679a4656-1d19-4f8e-b309-aaa26e9c2127)
![deletedateofbirth](https://github.com/CarlosZ96/first-look-at-sql/assets/125598459/5abdb151-2c5c-4dcf-946d-6fd4484c5af4)
![aftercaos](https://github.com/CarlosZ96/first-look-at-sql/assets/125598459/1bb0e29a-e819-4aa5-8309-3ece9c0f14cd)
![delete](https://github.com/CarlosZ96/first-look-at-sql/assets/125598459/cc641d24-1146-4166-bd8e-5274b6941d1d)
![setspeciescommit](https://github.com/CarlosZ96/first-look-at-sql/assets/125598459/c7dff193-e45b-4d86-be0b-73d0c616808f)
![setspecies](https://github.com/CarlosZ96/first-look-at-sql/assets/125598459/87bcc394-7b96-4621-b8d9-fcfbd6bd8c61)

CarlosZ96 committed:
Feature:

Learning to use transactions and filtering with operators

Strangeal commented on queries.sql:
```diff
@@ -23,4 +23,88 @@ SELECT * from animals WHERE name NOT like 'Gabumon';
+
+ROLLBACK TO speciesbefore;
+
+COMMIT;
```
- [x] Great work on your query to update the `species` column in the animal table! 🐾👏 However, I'd like to highlight a specific requirement from the project guidelines: it explicitly states that after updating the `species` column to `unspecified`, verify that the change was made successfully. Then roll back the change and `verify` that the `species` columns went back to the state before the transaction. 🔄

  - Upon reviewing your query, I noticed that you successfully updated the `species` column to `unspecified` within a transaction. However, it appears that you missed the verification steps outlined in the project guidelines. These verification steps are crucial to ensuring the integrity of the data and the correctness of your query.

    **To meet the project requirements, please consider adding the following verification step to your query:**

     1. After the update, verify that the change was made successfully. You can do this by executing a SELECT statement to confirm that the `species` column now contains the value `unspecified`. 🧐
     2. After the rollback, verify that the changes were made successfully. You can do this by executing a SELECT statement to confirm that the `species` column doesn't contain the value `unspecified`.
     
- [x] Kindly consider not using `SAVEPOINT` for this query as it's not really needed.

Adding these verification steps will not only ensure that your query meets the project requirements but also provide a robust and reliable way to confirm the correctness of your data manipulation.

Happy coding! 😊🚀🍀

CarlosZ96 replied:
- [x] Great work on your query to update the `species` column in the animal table! 🐾👏 However, I'd like to highlight a specific requirement from the project guidelines: it explicitly states that after updating the `species` column to `unspecified`, verify that the change was made successfully. Then roll back the change and `verify` that the `species` columns went back to the state before the transaction.:
![01](https://github.com/CarlosZ96/first-look-at-sql/assets/125598459/d089ef72-303a-489d-ae6a-94561cdfb9c8)

![02](https://github.com/CarlosZ96/first-look-at-sql/assets/125598459/fdd27ad1-844e-4597-857f-20e35dca4897)
After the rollback:

![03](https://github.com/CarlosZ96/first-look-at-sql/assets/125598459/f5cafcd2-0aa9-4c9e-a1bc-9b34d0521202)

- [x] Great work on crafting a transactional query to update the `species` column in the `animal` table! 🎉 However, as I highlighted in my previous comment regarding the significance of incorporating verification steps within transactional queries, it seems like you missed the verification step on this query as well. This step should be implemented before and after the `COMMIT` to ensure that the changes were successful following the update and commit. 👍

![04](https://github.com/CarlosZ96/first-look-at-sql/assets/125598459/a6bddfd0-d67b-4dd9-bf8c-d7493a30c702)

![05](https://github.com/CarlosZ96/first-look-at-sql/assets/125598459/36680ff1-7b4c-4cdd-8338-4fcd834ac78f)

![06](https://github.com/CarlosZ96/first-look-at-sql/assets/125598459/6facec4c-b7ad-4c58-bcbb-c8fa911bff42)

Strangeal commented on queries.sql:
```diff
@@ -23,4 +23,88 @@ SELECT * from animals WHERE name NOT like 'Gabumon';
+UPDATE animals
+SET species = 'pokemon' WHERE species IS NULL ;
+
+COMMIT;
```
- [x] Great work on crafting a transactional query to update the `species` column in the `animal` table! 🎉 However, as I highlighted in my previous comment regarding the significance of incorporating verification steps within transactional queries, it seems like you missed the verification step on this query as well. This step should be implemented before and after the `COMMIT` to ensure that the changes were successful following the update and commit. 👍

   - Adding these verification steps will help guarantee the integrity of your data and provide an extra layer of confidence in the transaction's success. 

Happy coding! 😊👩‍💻🚀

Strangeal commented on queries.sql:
```diff
@@ -23,4 +23,88 @@ SELECT * from animals WHERE name NOT like 'Gabumon';
+COMMIT;
+
+
+BEGIN;
+
+SAVEPOINT beforecaos;
+
+DELETE FROM animals;
+
+ROLLBACK TO beforecaos;
+
+COMMIT;
```
- [ ] Great work on crafting a transactional query to update the `animal` table! 🎉 However, as I highlighted in my previous comments regarding the significance of incorporating verification steps within transactional queries, it seems like you missed the verification step on this query as well. This step should be implemented before and after the `ROLLBACK` to ensure that the changes were successful. 👍

   - Adding these verification steps will help guarantee the integrity of your data and provide an extra layer of confidence in the transaction's success. 

Happy coding! 😊👩‍💻🚀

CarlosZ96 replied:
![07](https://github.com/CarlosZ96/first-look-at-sql/assets/125598459/4be96d1b-b8d6-4186-9990-cdae6d3a3dff)

![08](https://github.com/CarlosZ96/first-look-at-sql/assets/125598459/3daa9231-d30f-4a7d-9c63-d6f97a957d9e)

CarlosZ96 replied:
Corrected!

Strangeal commented on queries.sql:
```diff
@@ -23,4 +23,88 @@ SELECT * from animals WHERE name NOT like 'Gabumon';
+DELETE FROM animals WHERE  date_of_birth > '2022-01-01';
+
+ROLLBACK TO beforecaos;
+COMMIT;
```
- [ ] Great work on your database query implementation so far! 👏 However, I've noticed a couple of areas where your query doesn't align with the project requirements. Let's take a closer look at the specific steps outlined in the requirements:

<img width="740" alt="Screenshot 2023-10-05 at 23 24 47" src="https://github.com/CarlosZ96/first-look-at-sql/assets/107192076/a9665a7f-4fa7-4093-a3d1-770a69685678">

   - Upon reviewing these requirements, it appears that there are a few adjustments needed in your query. First, according to the requirements, you should execute a `DELETE` operation before creating the `SAVEPOINT`, but your current query doesn't follow this order. Additionally, your query lacks the necessary `UPDATE` steps both before and after the rollback operation.

To ensure that your queries produce accurate results and align with the project requirements, I recommend making the following changes:

1. Start by implementing the `DELETE` operation before creating the `SAVEPOINT`, as specified in the requirements.

2. Include the required `UPDATE` operations before and after the rollback step to maintain data integrity and meet the project's objectives.

These adjustments should bring your query in line with the project requirements and ensure it delivers the desired output. 
Happy coding! 😊🚀🍀

CarlosZ96 replied:
![09](https://github.com/CarlosZ96/first-look-at-sql/assets/125598459/e536dce6-997f-4443-850a-f953ab1e5cdd)
![10](https://github.com/CarlosZ96/first-look-at-sql/assets/125598459/ca583153-a03c-48d0-a259-911c80328ab8)
![11](https://github.com/CarlosZ96/first-look-at-sql/assets/125598459/895d92ec-893d-4dcf-a9e5-1457cba54e13)
![12](https://github.com/CarlosZ96/first-look-at-sql/assets/125598459/074a6035-6651-43ce-84c0-e7ea1e9cc217)
-Corrected!

Strangeal commented on queries.sql:
```diff
@@ -23,4 +23,88 @@ SELECT * from animals WHERE name NOT like 'Gabumon';
+WHERE species = 'digimon' AND date_of_birth BETWEEN '1990-01-01' AND '2000-12-31';
+SELECT SUM(escape_attempts) / COUNT(*) AS escape_attempts_average
+FROM animals
+WHERE species = 'pokemon' AND date_of_birth BETWEEN '1990-01-01' AND '2000-12-31';
```
- [ ] Great job on your query so far! 😄 However, there's a small adjustment needed to get the accurate result. Based on the data you've inserted into your database, the average number of escape attempts per animal type for those born between `1990` and `2000` is `3.000`.

   - To achieve this result, consider revising your query to ensure it accurately calculates the average number of escape attempts for animals born within that date range. Additionally, please update your screenshot with the correct output to reflect this refined query.

Happy coding! 😊🚀🍀

CarlosZ96 replied:
![15](https://github.com/CarlosZ96/first-look-at-sql/assets/125598459/9c5a5521-7cb1-41b0-81d5-59c1b95b612a)
-Corrected!

Strangeal commented on queries.sql:
```diff
@@ -23,4 +23,88 @@ SELECT * from animals WHERE name NOT like 'Gabumon';
+What is the average weight of animals?
+*/
+SELECT SUM(weight_kg) / COUNT(*) AS weight_kg_average
+FROM animals;
```
- [ ] Great job on your query so far! 😄 However, there's a small adjustment needed to get an accurate result. Based on the data you've inserted into your database, the average weight of animals should be `15.XXX`. 

  - To achieve this result, consider revising your query to ensure it accurately calculates the average weight of animals. Additionally, please update your screenshot with the correct output to reflect this refined query.
  
Happy coding! 😊🚀🍀

CarlosZ96 replied:
![13](https://github.com/CarlosZ96/first-look-at-sql/assets/125598459/2e3026f9-24ea-4916-b920-8a8aaac733f9)

Strangeal commented on queries.sql:
```diff
@@ -23,4 +23,88 @@ SELECT * from animals WHERE name NOT like 'Gabumon';
+SELECT COUNT(escape_attempts) FROM animals WHERE neutered = TRUE;
+SELECT COUNT(escape_attempts) FROM animals WHERE neutered = FALSE;
+/*
+What is the minimum and maximum weight of each type of animal?
+*/
+SELECT weight_kg
+FROM animals
+ORDER BY weight_kg ASC LIMIT 1;
+SELECT weight_kg
+FROM animals
+ORDER BY weight_kg DESC LIMIT 1;
```
- [ ] Great job on your query so far! 😄 However, to ensure an accurate result, a small adjustment is needed. Based on the data you've inserted into your database, you should consider selecting the `species` column from the `animal` table instead of `weight_kg` to determine the `max` and `min` weight of each type of animal.

I recommend rewriting your query to achieve the desired output. Here's a revised version:

```sql
SELECT
    species,
    MAX(weight_kg) AS max_weight,
    MIN(weight_kg) AS min_weight
FROM
    animal
GROUP BY
    species;
```

This query will correctly provide you with the maximum and minimum weights for each type of animal in your database.

Happy coding! 😊👍

CarlosZ96 replied:
![14](https://github.com/CarlosZ96/first-look-at-sql/assets/125598459/68033590-0ae5-47c0-9d29-d9933123740e)
-Corrected!

Strangeal reviewed (changes requested):
Hi @CarlosZ96, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!


![](https://media.giphy.com/media/v1.Y2lkPTc5MGI3NjExOGYyYWNkMjRkY2JjYjg1MjY1M2NjYTNkNWNiNTQwN2EyZGZmNWM5YSZlcD12MV9pbnRlcm5hbF9naWZzX2dpZklkJmN0PWc/m3864rBwwBTKMAbpn3/giphy.gif)

## Highlights
- Good **PR** title ✅ 
- Descriptive *PR* description ✅ 
- Implement the features ✅ 
- Professional `README` template ✅ 
- Meaningful **Github** commits ✅ 
- Well structured project ✅ 

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

CarlosZ96 committed:
Update queries.sql

the requested changes are made in the PR the species column is updated and returned to it without the use of savepoint screenshots attached to the comment in the PR

CarlosZ96 committed:
Update queries.sql

The requested changes were made by updating the species column and checking as requested, leaving the screenshots a comment in the PR

CarlosZ96 committed:
Update queries.sql

CarlosZ96 committed:
Update queries.sql

minor fix

CarlosZ96 committed:
Update queries.sql

ademibowale reviewed (approved):
Hi @CarlosZ96  , 


Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## To Highlight  
- Descriptive pull request ✔️
- Everything that populates tables with data should be saved in the data.sql file✔️
- Everything that is a query that helps to analyze data should be saved in the queries.sql file ✔️
- Descriptive commit messages ✔️ 

![](https://media.giphy.com/media/mCIjCgs3nWQWfJZvPA/giphy-downsized-large.gif)




- [x]  _Please check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me on( @Adebowale ) in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
